google/cloud/bigtable/examples/table_admin_snippets.cc

```diff
@@ -0,0 +1,351 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! [all code]
+
+//! [bigtable includes]
+#include "google/cloud/bigtable/table.h"
+#include "google/cloud/bigtable/table_admin.h"
+//! [bigtable includes]
+#include <google/protobuf/text_format.h>
+#include <deque>
+#include <list>
+#include <sstream>
+
+namespace {
+/**
+ * The key used for ReadRow(), ReadModifyWrite(), CheckAndMutate.
+ *
+ * Using the same key makes it possible for the user to see the effect of
+ * the different APIs on one row.
+ */
+
+struct Usage {
+  std::string msg;
+};
+
+char const* ConsumeArg(int& argc, char* argv[]) {
+  if (argc < 2) {
+    return nullptr;
+  }
+  char const* result = argv[1];
+  std::copy(argv + 2, argv + argc, argv + 1);
+  argc--;
+  return result;
+}
+
+std::string command_usage;
+
+void PrintUsage(int argc, char* argv[], std::string const& msg) {
+  std::string const cmd = argv[0];
+  auto last_slash = std::string(cmd).find_last_of('/');
+  auto program = cmd.substr(last_slash + 1);
+  std::cerr << msg << "\nUsage: " << program << " <command> [arguments]\n\n"
+            << "Commands:\n"
+            << command_usage << std::endl;
+}
+
+//! [create table]
```

Comment:
In a future PR we should move these markers inside the function, to exclude the noise checking for the right number of parameters and the `ConsumeArg` stuff.  This would become:

```C++
void CreateTable(google::cloud::bigtable::TableAdmin admin, int argc,
                 char* argv[]) {
  if (argc != 2) {
    throw Usage{"create-table: <project-id> <instanse-id> <table-id>"};
  }
  std::string const table_id = ConsumeArg(argc, argv);

  //! [create table]
  [](google::cloud::bigtable::TableAdmin admin, std::string table_id) {
   auto schema = admin.CreateTable(
      table_id,
      google::cloud::bigtable::TableConfig(
          // The schema is hardcoded in these examples to keep them simple.
          {{"fam", google::cloud::bigtable::GcRule::MaxNumVersions(10)},
           {"foo",
            google::cloud::bigtable::GcRule::MaxAge(std::chrono::hours(72))}},
          {}));
  }
  //! [create table]
  (std::move(admin), table_id);
}
```
That way the documentation only shows the interesting bits, and excludes all the parsing noise.